Decouple the calendar, FAQ and news bundles from the comments bundle

zoglo opened the pull request:
**Feature**

- Implements #3510 

**Description**

- This PR removes the hard dependency of the comment-bunndle within faq, news and calendar by using the PaletteManipulator
- All language files have been separated as well

**Following things haven't been removed**

- Comment implementation within the respecitve Modules
- Comment implementation within Models
- Comment implementation within Templates

**Related**
- #3510
- https://github.com/contao/contao/pull/2983#issuecomment-839812547
- #3508

_____
*Disclaimer*
> The amount of files and lines changed is due to me actually moving the translations of multiple languages into the comments-bundle

zoglo committed:
Rework conditional comment / news bundle fields

- See #3510

zoglo committed:
Rework conditional comment / faq bundle fields

- See #3510

zoglo committed:
Rework conditional comment / calendar bundle fields

- See #3510

zoglo committed:
CS

zoglo committed:
CS #2

zoglo committed:
Revert comment / calendar translations

- english only needed for transifex

zoglo committed:
Revert comment / faq translations

- english only needed for transifex

zoglo committed:
Revert comment / news translations

- english only needed for transifex

zoglo committed:
Revert auto linting eol

zoglo committed:
CS

zoglo committed:
Revert editorconfig

aschempp reviewed (changes requested):
Great improvement! There is one major problem though. Installing the comments bundle will now define fields on e.g. `tl_calendar_events`, even if the calendar bundle is not installed. We should either check for that in the DCA (which is kinda hacky) or dynamically add the fields using a `loadDataContainer` callback.

zoglo commented:
> Great improvement! There is one major problem though. Installing the comments bundle will now define fields on e.g. `tl_calendar_events`, even if the calendar bundle is not installed. We should either check for that in the DCA (which is kinda hacky) or dynamically add the fields using a `loadDataContainer` callback.

I'll get onto it later this evening :)

zoglo committed:
Conditionally load fields via DataContainerListener

- use loadDataContainer Hook
- universal size
  - notify is now varchar(32) for faq and news
  - sortOrder is now varchar(32) for faq

zoglo committed:
CS

zoglo committed:
CS

zoglo committed:
CS

- satisfy linter

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

aschempp commented on comments-bundle/src/ContaoManager/Plugin.php:
```diff
@@ -27,7 +27,7 @@ public function getBundles(ParserInterface $parser): array
         return [
             BundleConfig::create(ContaoCommentsBundle::class)
-                ->setLoadAfter([ContaoCoreBundle::class])
+                ->setLoadAfter([ContaoCoreBundle::class, 'calendar', 'faq', 'news'])
```
This should use the bundle names, not the legacy module names.

zoglo replied:
@aschempp using the bundle names would need use statements which would ultimately lead to have the bundles installed again 🤔?

aschempp replied:
no, use statements are not package requirements. A class name is just a string internally, the class does not have to be present.

aschempp commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,176 @@
+            case 'tl_news':
+            case 'tl_calendar_events':
+                PaletteManipulator::create()
+                    ->addLegend('comments_legend', 'protected_legend', PaletteManipulator::POSITION_AFTER, true)
```
```suggestion
                    ->addLegend('comments_legend', null, PaletteManipulator::POSITION_APPEND, true)
```

Do we really need to specify the parent legend? Or could we maybe just append to the end and not have to distinguish between the tables? Not sure my code works as is! 😅

aschempp commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,176 @@
+
+            case 'tl_faq':
+                PaletteManipulator::create()
+                    ->addLegend('expert_legend', 'publish_legend', PaletteManipulator::POSITION_BEFORE, true)
```
if the legend already exists, the palette manipulator should simply ignore this (or maybe change the order 🤔). So maybe we could apply the same call to all tables?

zoglo replied:
The expert legend does not exist in the faq table, hence why I am handling most of it this way. We could add it within the faq-bundle without a field so it won't be shown, wdyt?

aschempp replied:
That's not what I mean. If you use this code in the news and calendar bundle as well, it should simply do nothing because the legend already exists. So no switch-case would be necessary.

zoglo replied:
I checked this again, the problem ist that news and calendar_events have 3 more palettes that this option has to be added:
```php
PaletteManipulator::create()
    ->addField(['noComments'], 'expert_legend', PaletteManipulator::POSITION_APPEND)
    ->applyToPalette('default', $table)
    ->applyToPalette('internal', $table)
    ->applyToPalette('article', $table)
    ->applyToPalette('external', $table)
;
```

Whilst the faq bundle only has one palette (default). The only way to circumvent this would be changing the PaletteNotFoundException to not be thrown 🤔 ?

aschempp commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,176 @@
+            case 'tl_faq':
+                PaletteManipulator::create()
+                    ->addLegend('expert_legend', 'publish_legend', PaletteManipulator::POSITION_BEFORE, true)
+                    ->addField(['noComments'], 'expert_legend', PaletteManipulator::POSITION_APPEND)
```
```suggestion
                    ->addField('noComments', 'expert_legend', PaletteManipulator::POSITION_APPEND)
```

Just a minor nitpick and could also stay as-is. Applies to other places as well.

aschempp commented on comments-bundle/contao/dca/tl_module.php:
```diff
@@ -69,3 +70,32 @@
+
+$bundles = System::getContainer()->getParameter('kernel.bundles');
+
+if (isset($bundles['ContaoNewsBundle']))
```
I would prefer to add this to a/the listener as well, the DCA code should not contain conditional stuff (e.g. something that would not be possible in YAML) imho. I know that's not always the case atm, but kinda best practice 😅

aschempp commented on comments-bundle/src/DependencyInjection/ContaoCommentsExtension.php:
```diff
@@ -22,6 +22,7 @@ class ContaoCommentsExtension extends Extension
     public function load(array $configs, ContainerBuilder $container): void
     {
         $loader = new YamlFileLoader($container, new FileLocator(__DIR__.'/../../config'));
+        $loader->load('listener.yaml');
```
not sure we really need a new `listener.yaml` file just for one service? @contao/developers wdyt?

zoglo replied:
@aschempp anything regarding this? 🤔

fritzmg replied:
I think I agree. The smaller bundles should just have one `services.yaml` containing all the service definitions, no need to organize them (as there won't be that many).

aschempp reviewed (changes requested):
Very good! This shows how much duplicate code there was until now. Just a few minor things to fix for me 😎

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

zoglo committed:
Improve DataContainerListener and more

- do not use legacy module names anymore
- unionize faq, comments and news withing palette manipulators
- CS

zoglo committed:
CS

- remove unused Use statement

fritzmg commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,160 @@
+
+        switch ($table) {
+            case 'tl_module':
+                self::applyModuleFields();
```
What is the advantage of `self::` over `$this->`? I don't think we are using `self::` for non-static method calls anywhere else.

zoglo replied:
~~Was done out of habit due to referring to the class (and the methods technically being able to be static). 
I can change that.~~

Will be changed due to using %kernel.bundles%

fritzmg commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,160 @@
+
+    private function applyModuleFields(): void
+    {
+        $bundles = System::getContainer()->getParameter('kernel.bundles');
```
I don't think we should use `System::getContainer()` in regular services. Why not inject `%kernel.bundles%`?

zoglo committed:
Updates to DataContainerListener

- register listener within services.yaml (and remove listener.yaml)
- get ``kernel.bundles`` via argument
- properly call methods (object and not class)

aschempp commented on comments-bundle/src/ContaoManager/Plugin.php:
```diff
@@ -27,8 +30,12 @@ public function getBundles(ParserInterface $parser): array
+                    ContaoCalendarBundle::class,
+                    ContaoFaqBundle::class,
+                    ContaoNewsBundle::class],
+                )->setReplace(['comments']),
```
```suggestion
                    ContaoNewsBundle::class,
                ])->setReplace(['comments']),
```

Not sure why the code style checker does not mention this 🤷

aschempp commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,161 @@
+        if ('tl_news' === $table || 'tl_calendar_events' === $table) {
+            $GLOBALS['TL_DCA'][$table]['fields']['noComments']['eval'] = ['tl_class' => 'w50 m12'];
+
+            $pm->applyToPalette('internal', $table)
```
```suggestion
            $pm
                ->applyToPalette('internal', $table)
```

aschempp reviewed (changes requested):
Perfect! Just found two code style issues 🙃

zoglo committed:
CS

zoglo commented:
> Perfect! Just found two code style issues 🙃

Did your changes in the latest commit. I ran ecs locally but it did skip this one 🤷‍♂️

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

zoglo committed:
Add plugin test

zoglo committed:
Merge remote-tracking branch 'origin/feature/comments-conditional-fields' into feature/comments-conditional-fields

zoglo committed:
Update test name

aschempp reviewed (dismissed):
🎉

Toflar reviewed (dismissed):
Wohoo, thanks so much for taking the time to finally separate this! Not a feature a user will ever see (other than not having useless fields in the table anymore 🥳 ) but a very valuable cleanup-contribution! ❤️

Toflar commented:
We'll probably need a test for the new `CommentsConditionalFieldsListener` though.

zoglo committed:
Add CommentsConditionalFieldsListenerTest

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

zoglo committed:
Update tests

zoglo committed:
Unset TL_LANG after test

zoglo committed:
CS

Toflar reviewed (dismissed):
Perfect!

zoglo committed:
Update tests (100% code cov)

leofeyer committed:
Merge branch '5.x' into feature/comments-conditional-fields

leofeyer commented on comments-bundle/src/EventListener/DataContainer/CommentsConditionalFieldsListener.php:
```diff
@@ -0,0 +1,162 @@
+use Contao\CoreBundle\DependencyInjection\Attribute\AsHook;
+
+#[AsHook('loadDataContainer')]
+class CommentsConditionalFieldsListener
```
Should we rename this to `AddCommentFieldsListener`?

zoglo replied:
> Should we rename this to `AddCommentFieldsListener`?

Same for the test, will change that later.

zoglo replied:
Did rename it and changed it within services.yaml as well @leofeyer

leofeyer committed:
CS

zoglo committed:
Merge branch '5.x' into feature/comments-conditional-fields

zoglo committed:
Rename `CommentsConditionalFields` to `AddCommentFields`

leofeyer reviewed (approved):
Perfect. Than you very much @zoglo.
